File: go-leetcode/backtracking/216.组合总和III.go
package leetcode

func combinationSum3(k int, n int) [][]int {
	res := [][]int{}
	path := []int{}

	var backtrack func(start int, sum int)
	backtrack = func(start int, sum int) {
		if len(path) == k {
			if sum == n {
				temp := make([]int, k)
				copy(temp, path)
				res = append(res, temp)
			}
			return
		}
		// 剪枝：最多到 9 - (k - len(path)) + 1
		for i := start; i <= 9-(k-len(path))+1; i++ {
			path = append(path, i)
			backtrack(i+1, sum+i)
			path = path[:len(path)-1]
		}
	}

	backtrack(1, 0)
	return res
}

File: python-leetcode/backtracking/216.组合总和III.py
from typing import List


class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        res: List[List[int]] = []
        path: List[int] = []

        def backtrack(start: int, current_sum: int):
            if len(path) == k:
                if current_sum == n:
                    res.append(path[:])
                return

            # 剪枝：最多到 9 - (k - len(path)) + 1
            for i in range(start, 9 - (k - len(path)) + 2):
                path.append(i)
                backtrack(i + 1, current_sum + i)
                path.pop()

        backtrack(1, 0)
        return res
